File: backend/services/redisService.js
import Redis from 'ioredis';
import dotenv from 'dotenv';
dotenv.config();

const REDIS_URL = process.env.REDIS_URL || 'redis://localhost:6379';

class CacheService {
  constructor() {
    this.memoryStore = new Map();
    this.memoryTimeouts = new Map();
    this.isRedisConnected = false;
    this.redisClient = null;

    this.init();
  }

  init() {
    try {
      this.redisClient = new Redis(REDIS_URL, {
        maxRetriesPerRequest: 1,
        retryStrategy: () => null, // Don't crash or loop endlessly if Redis is offline
        connectTimeout: 2000,
        lazyConnect: true
      });

      this.redisClient.connect()
        .then(() => {
          this.isRedisConnected = true;
          console.log(`[Redis] 🟢 Connected to Redis at ${REDIS_URL}`);
        })
        .catch((err) => {
          this.isRedisConnected = false;
          console.log(`[Redis] ⚠️ Redis server not reachable at ${REDIS_URL}. Using high-speed in-memory cache fallback.`);
        });

      this.redisClient.on('error', (err) => {
        if (this.isRedisConnected) {
          console.warn('[Redis] Connection lost, switching to in-memory fallback.');
        }
        this.isRedisConnected = false;
      });
    } catch (e) {
      this.isRedisConnected = false;
      console.log('[Redis] Falling back to in-memory cache store.');
    }
  }

  async get(key) {
    if (this.isRedisConnected && this.redisClient) {
      try {
        const data = await this.redisClient.get(key);
        return data ? JSON.parse(data) : null;
      } catch (err) {
        this.isRedisConnected = false;
      }
    }
    const val = this.memoryStore.get(key);
    return val !== undefined ? JSON.parse(val) : null;
  }

  async set(key, value, ttlSeconds = 3600) {
    const serialized = JSON.stringify(value);

    if (this.isRedisConnected && this.redisClient) {
      try {
        if (ttlSeconds) {
          await this.redisClient.set(key, serialized, 'EX', ttlSeconds);
        } else {
          await this.redisClient.set(key, serialized);
        }
        return true;
      } catch (err) {
        this.isRedisConnected = false;
      }
    }

    // In-memory fallback
    this.memoryStore.set(key, serialized);
    if (this.memoryTimeouts.has(key)) {
      clearTimeout(this.memoryTimeouts.get(key));
    }
    if (ttlSeconds) {
      const timeout = setTimeout(() => {
        this.memoryStore.delete(key);
        this.memoryTimeouts.delete(key);
      }, ttlSeconds * 1000);
      this.memoryTimeouts.set(key, timeout);
    }
    return true;
  }

  async del(key) {
    if (this.isRedisConnected && this.redisClient) {
      try {
        await this.redisClient.del(key);
      } catch (e) {}
    }
    this.memoryStore.delete(key);
    if (this.memoryTimeouts.has(key)) {
      clearTimeout(this.memoryTimeouts.get(key));
      this.memoryTimeouts.delete(key);
    }
  }

  async clearQueries() {
    for (const key of this.memoryStore.keys()) {
      if (key.startsWith('query:')) {
        this.memoryStore.delete(key);
      }
    }
  }

  getStatus() {
    return {
      connected: this.isRedisConnected,
      mode: this.isRedisConnected ? 'redis' : 'in-memory-fallback',
      entriesCached: this.memoryStore.size
    };
  }
}

export const cacheService = new CacheService();

File: backend/services/redis_service.py
import os
import time
import json
import logging
from typing import Any, Optional

logger = logging.getLogger("redis_service")

class CacheService:
    def __init__(self):
        self.redis_client = None
        self.is_redis_available = False
        self.memory_store = {}
        self.memory_expiries = {}
        self._connect()

    def _connect(self):
        redis_host = os.getenv("REDIS_HOST", "127.0.0.1")
        redis_port = int(os.getenv("REDIS_PORT", "6379"))
        redis_url = os.getenv("REDIS_URL")

        try:
            import redis
            if redis_url:
                self.redis_client = redis.Redis.from_url(redis_url, socket_connect_timeout=1, socket_timeout=1)
            else:
                self.redis_client = redis.Redis(host=redis_host, port=redis_port, socket_connect_timeout=1, socket_timeout=1)
            
            self.redis_client.ping()
            self.is_redis_available = True
            logger.info("Connected to Redis cache layer.")
        except Exception:
            self.is_redis_available = False
            self.redis_client = None
            logger.info("Redis not reachable. Operating with in-memory caching.")

    def get_status(self) -> dict:
        return {
            "connected": self.is_redis_available,
            "type": "redis" if self.is_redis_available else "in-memory"
        }

    async def get(self, key: str) -> Optional[Any]:
        if self.is_redis_available and self.redis_client:
            try:
                raw = self.redis_client.get(key)
                if raw:
                    return json.loads(raw)
            except Exception as e:
                logger.warning(f"Redis get error: {e}")
        
        # Memory fallback
        if key in self.memory_store:
            expiry = self.memory_expiries.get(key)
            if expiry and time.time() > expiry:
                del self.memory_store[key]
                del self.memory_expiries[key]
                return None
            return self.memory_store[key]
        return None

    async def set(self, key: str, value: Any, ttl_seconds: int = 86400) -> bool:
        if self.is_redis_available and self.redis_client:
            try:
                payload = json.dumps(value)
                self.redis_client.setex(key, ttl_seconds, payload)
                return True
            except Exception as e:
                logger.warning(f"Redis set error: {e}")

        # Memory fallback
        self.memory_store[key] = value
        if ttl_seconds:
            self.memory_expiries[key] = time.time() + ttl_seconds
        return True

    async def delete(self, key: str) -> bool:
        if self.is_redis_available and self.redis_client:
            try:
                self.redis_client.delete(key)
            except Exception:
                pass
        self.memory_store.pop(key, None)
        self.memory_expiries.pop(key, None)
        return True

cache_service = CacheService()
